internal/controllers/synthesis/lifecycle.go

```diff
@@ -317,6 +320,30 @@ func shouldDeletePod(logger logr.Logger, comp *apiv1.Composition, syn *apiv1.Syn
 	return logger, nil, false
 }
 
+// deletePod deletes one Pod associated to the given comp unconditionally.
+// Should only be used when the composition no longer exists.
+func (c *podLifecycleController) deletePod(ctx context.Context, comp types.NamespacedName) error {
+	logger := logr.FromContextOrDiscard(ctx)
+	pods := &corev1.PodList{}
+	if err := c.client.List(ctx, pods, client.InNamespace(c.config.PodNamespace), client.MatchingFields{
+		manager.IdxPodsByComposition: manager.PodByCompIdxValueFromNamespacedName(comp),
+	}); err != nil {
+		return fmt.Errorf("listing Pods: %w", err)
+	}
+	for _, pod := range pods.Items {
+		if pod.DeletionTimestamp != nil {
+			continue
+		}
+		err := c.client.Delete(ctx, &pod)
+		if client.IgnoreNotFound(err) != nil {
+			return fmt.Errorf("deleting Pod %s: %w", pod.Name, err)
+		}
+		logger.V(0).Info("deleted synthesizer pod", "podName", pod.Name, "reason", "CompositionDoesNotExist")
+		return nil
```

Comment:
why delete one pod only? should we delete all synthesizer pod if the comp doesn't exist?

Reply:
This controller watches both Pods and Compositions, it should only do one write of the most granular resource per controller tick (the Pod in this case). The deletion of one Pod (either successful or unsuccessful) will trigger another controller tick and thus another deletion until we're done